### PR TITLE
Feature/feedme

### PR DESCRIPTION
**Pull request by christoph-holter**
Adding feedme command to get info from [Fabrik](http://www.diefabrik.co.at/mittagsmenue/index.html) and [Ernis](http://www.ernis.at/).

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,10 +1,11 @@
 {
   "name": "hubot",
-  "version": "0.0.0",
+  "version": "1.0.0",
```
Comment (lorenzleutgeb) [inline package.json]: Going 1.0 seems like a big step to me. How about applying the suggestions postulated in [Semantic Versioning 2.0.0](http://semver.org/#how-should-i-deal-with-revisions-in-the-0yz-initial-development-phase), e.g. use version 0.1.0 for this PR?

File: package.json [hunk 2]
```diff
@@ -1,10 +1,11 @@
...
   "private": true,
   "author": "Lorenz Leutgeb <lorenz.leutgeb@catalysts.cc>",
   "description": "A simple helpful robot for Catalysts",
   "dependencies": {
+    "datejs": "^1.0.0-rc3",
```
Comment (lorenzleutgeb) [inline package.json]: Depends on a release candidate. Scary. Needs justification.

File: scripts/feedme.coffee [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+fetch = (target) ->
+  return urllib.request(target).data
+
+
+class Fabrik
+  target : "http://www.diefabrik.co.at/mittagsmenue/index.html"
+  holidayMagic : "urlaub"
+  errors:
+	  closed : "fabrik is closed"
```
Comment (lorenzleutgeb) [inline scripts/feedme.coffee]: Formatting of lines 28 through 32 is inconsistent with formatting of line 25 and 26.

File: scripts/feedme.coffee [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+	  menuFromPast :"the menu is outdated"
+	  menuFromFuture :"the menu is from the future"
+	  resting :"fabrik is on a day off"
+	  holiday :"fabrik is probably on holiday, fall back to manual check"
+
+  constructor: (@robot) ->
+
+  checkHoliday: (text) =>
+    return true if text.toLowerCase().indexOf(@holidayMagic) > 0
```
Comment (lorenzleutgeb) [inline scripts/feedme.coffee]: Assuming `return true if x` returns something falsy if `not x`, you can omit this altogether and write `x`.

File: scripts/feedme.coffee [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+  parseDates: ($) ->
+    docDate = $("h2:eq(0)").html()
+    return /(\d{2}\.\d{2}\.\d{4}) bis (\d{2}\.\d{2}\.\d{4})/.exec(docDate)
+
+  extractMeal: ($, day) ->
+
+    # Saturday & Sunday
+    return @errors.closed if day == 6 | day == 0
```
Comment (lorenzleutgeb) [inline scripts/feedme.coffee]: Are you sure that `|` is logic "or" just like `or`? [This](https://github.com/jashkenas/coffeescript/issues/281) does not look like a logic operator.

File: scripts/feedme.coffee [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    # Ruhetag
+    return @errors.resting if menu.toLowerCase() == 'ruhetag'
+    # all good :)
+    return menu
+
+  getMenue: () =>
+    now = new Date()
+    day = now.getDay()
+    lastCheck = new Date(@robot.brain.get "feedme.fabrik.lastCheck")
+    lastCheck = Date.parseExact("01.01.1970", "d.M.yyyy") if not lastCheck
```
Comment (lorenzleutgeb) [inline scripts/feedme.coffee]: Could be `lastCheck = new Date(@robot.brain.get("feedme.fabrik.lastCheck") || 0)`.

File: scripts/feedme.coffee [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    docDate = $("h2:eq(0)").html()
+    return /(\d{2}\.\d{2}\.\d{4}) bis (\d{2}\.\d{2}\.\d{4})/.exec(docDate)
+
+  extractMeal: ($, day) ->
+
+    # Saturday & Sunday
+    return @errors.closed if day == 6 | day == 0
+    menu = $(".contenttable .tr-#{(day-1)*2} .td-2").html()
+    # Ruhetag
+    return @errors.resting if menu.toLowerCase() == 'ruhetag'
```
Comment (lorenzleutgeb) [inline scripts/feedme.coffee]: Could be `if menu.toLowerCase() is 'ruhetag' then @errors.resting else menu` without any `return`.

File: scripts/feedme.coffee [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+    # all good :)
+    return menu
+
+  getMenue: () =>
```
Comment (lorenzleutgeb) [inline scripts/feedme.coffee]: ["menue" should be "menu".](dict.cc/?s=menue)

File: scripts/feedme.coffee [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+    return @errors.resting if menu.toLowerCase() == 'ruhetag'
+    # all good :)
+    return menu
+
+  getMenu: () =>
+    now = new Date()
+    day = now.getDay()
+    lastCheck = new Date(@robot.brain.get "feedme.fabrik.lastCheck" or 0)
+    # only check once per day, otherwise we're good to go
+    if getDateDiff(now, lastCheck, DAY) > 0
```
Comment (lorenzleutgeb) [inline scripts/feedme.coffee]: I think an [early return](http://programmers.stackexchange.com/questions/18454) would be nice here. All the code below would need one level of indentation less.

File: scripts/feedme.coffee [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+    menu = $(".contenttable .tr-#{(day-1)*2} .td-2").html()
+    # Ruhetag
+    return @errors.resting if menu.toLowerCase() == 'ruhetag'
+    # all good :)
+    return menu
+
+  getMenu: () =>
+    now = new Date()
+    day = now.getDay()
+    lastCheck = new Date(@robot.brain.get "feedme.fabrik.lastCheck" or 0)
```
Comment (lorenzleutgeb) [inline scripts/feedme.coffee]: The `or` operator takes precedence over function calls, so this is missing parentheses.

File: scripts/feedme.coffee (hunk 3 above)
Comment (flowlo) [inline scripts/feedme.coffee]: Still valid.

File: scripts/feedme.coffee [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+
+fetch = (target) ->
+  return urllib.request(target).data
+
+
+class Fabrik
+  target : "http://www.diefabrik.co.at/mittagsmenue/index.html"
+  holidayMagic : "urlaub"
+  errors:
+    closed : "fabrik is closed"
```
Comment (lorenzleutgeb) [inline scripts/feedme.coffee:24]: Formatting of lines 28 through 32 is inconsistent with formatting of line 25 and 26.

File: scripts/feedme.coffee [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+    menuFromPast :"the menu is outdated"
+    menuFromFuture :"the menu is from the future"
+    resting :"fabrik is on a day off"
+    holiday :"fabrik is probably on holiday, fall back to manual check"
+
+  constructor: (@robot) ->
+
+  checkHoliday: (text) =>
+    return text.toLowerCase().indexOf(@holidayMagic) > 0
```
Comment (lorenzleutgeb) [inline scripts/feedme.coffee]: `return` can be omitted.

File: scripts/feedme.coffee [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+
+  getMenu: () =>
+    now = new Date()
+    day = now.getDay()
+    lastCheck = new Date(@robot.brain.get "feedme.fabrik.lastCheck" or 0)
+    # only check once per day, otherwise we're good to go
+    if getDateDiff(now, lastCheck, DAY) > 0
+      rawbody = fetch(@target)
+
+      return @errors.holiday if @checkHoliday(rawbody.toString())
```
Comment (lorenzleutgeb) [inline scripts/feedme.coffee]: I think `return` can be omitted here as well.

File: scripts/feedme.coffee [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+module.exports = (robot) ->
+
+  fabrik = new Fabrik(robot)
+  ernis = new Ernis(robot)
+  robot.respond /feedme/i, (res) ->
+    fabrik.getMenu()
+    ernis.getMenu()
+    fabrikMenu = robot.brain.get "feedme.fabrik.save"
+    ernisMenu = robot.brain.get "feedme.ernis.save"
+    res.send "Heutiges Mittagsmenü: \nFabrik: \n#{fabrikMenu}\n\nErnis: \n#{ernisMenu}"
```
Comment (lorenzleutgeb) [inline scripts/feedme.coffee]: Should be `"...Erni's..."`.

File: scripts/feedme.coffee [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+
+    lastCheck = new Date(@robot.brain.get "feedme.ernis.lastCheck" or 0)
+    # only check once per day, otherwise we're good to go
+    if getDateDiff(now, lastCheck, DAY) > 0
+      rawbody = fetch(@target)
+
+      # hack to include jQuery
+      $ = require("jquery")(jsdom.jsdom(rawbody).defaultView)
+      # ernis closed
+      if day > 0 & day < 6
```
Comment (lorenzleutgeb) [inline scripts/feedme.coffee]: Not sure about `&` again, `and` is safe.

**Pull request merged**